src/v/kafka/server/handlers/describe_configs.cc

```diff
@@ -77,12 +79,124 @@ static ss::sstring describe_as_string(const T& t) {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+// property_config_type maps the datatype for a config property to
+// describe_configs_type. Currently class_type and password are not used in
+// Redpanda so we do not include checks for those types. You may find a similar
+// mapping in Apache Kafka at
+// https://github.com/apache/kafka/blob/be032735b39360df1a6de1a7feea8b4336e5bcc0/core/src/main/scala/kafka/server/ConfigHelper.scala
+// Type sizes are based on Java and taken from
+// https://docs.oracle.com/javase/tutorial/java/nutsandbolts/datatypes.html
+template<typename T>
+consteval describe_configs_type property_config_type() {
+    if constexpr (reflection::is_std_optional<T>) {
+        return property_config_type<typename T::value_type>();
+    } else if constexpr (reflection::is_tristate<T>) {
+        return property_config_type<typename T::value_type>();
+    } else if constexpr (std::is_same_v<T, bool>) {
+        return describe_configs_type::boolean;
+    } else if constexpr (std::is_same_v<T, ss::sstring>) {
+        return describe_configs_type::string;
+    } else if constexpr (std::is_integral_v<T>) {
+        if constexpr (is_short<T>()) {
+            // Types such as "signed char" will map here
+            return describe_configs_type::short_type;
+        } else if constexpr (is_int<T>()) {
+            return describe_configs_type::int_type;
+        } else if constexpr (is_long<T>()) {
+            return describe_configs_type::long_type;
+        } else {
+            return describe_configs_type::int_type;
+        }
+    } else if constexpr (std::is_floating_point_v<T>) {
+        return describe_configs_type::double_type;
+    } else if constexpr (reflection::is_std_vector<T>) {
+        return describe_configs_type::list;
+    } else if constexpr (std::is_same_v<T, model::compression>) {
+        return describe_configs_type::string;
+    } else if constexpr (std::is_same_v<T, model::cleanup_policy_bitflags>) {
+        return describe_configs_type::string;
+    } else if constexpr (std::is_same_v<T, std::chrono::seconds>) {
+        // Long type since seconds is atleast a 35-bit signed integral
+        // https://en.cppreference.com/w/cpp/chrono/duration
+        return describe_configs_type::long_type;
+    } else if constexpr (std::is_same_v<T, std::chrono::milliseconds>) {
+        // Long type since milliseconds is atleast a 45-bit signed integral
+        // https://en.cppreference.com/w/cpp/chrono/duration
+        return describe_configs_type::long_type;
+    } else if constexpr (std::is_same_v<T, model::timestamp_type>) {
+        return describe_configs_type::string;
+    } else if constexpr (std::is_same_v<T, config::data_directory_path>) {
+        return describe_configs_type::string;
+    } else if constexpr (std::is_same_v<T, v8_engine::data_policy>) {
+        return describe_configs_type::string;
+    } else {
+        static_assert(
```

Comment:
If moved outside of the `else` this is a guaranteed catch all in the case no `return` statement was ever hit

Reply:
Unfortunately, moving the static assert outside made all types fail the type check. I think this has something to do with constexpr. Let's keep the current style with the `else` for now since this follows the same style that's in `src/v/config/property.h`

Reply:
FWIW, `signed char` maps to `int_type` with the new type_trait checks so I think we're good here.

Reply:
Ah it makes sense why that would fail now that i think about it, because that is a code path that will always be evaluated at compile time.

What you have here is perfect then, since now after you most recent changes , in each branch there is a guaranteed return statement